File: FutureVoice/Services/MicPreference.swift
import AVFoundation
import Foundation

/// Which mic records the learner **when a Bluetooth earphone is connected**.
///
/// With nothing connected there is no choice to make — iOS uses the built-in
/// mic either way — so everything here is scoped to the Bluetooth case.
///
/// Neither is simply "better", which is why this is asked rather than
/// inferred:
///
/// - **the phone's mic is the better microphone.** Wideband, multi-mic
///   beamforming, no compression — everything the earphone gives up. Its one
///   weakness is that it is wherever the phone is: on a desk, in a bag,
///   face-down, in a pocket, which the app cannot see.
/// - **the earphone's mic has the better position.** It runs over HFP, which
///   narrows the link to telephone bandwidth (~16 kHz mSBC on modern
///   earphones) and adds the earphone's own aggressive processing — but it is
///   at the learner's mouth regardless of where the phone ended up, and that
///   is what carries it in noise and at distance.
///
/// So the default is the worn mic: it is the one that cannot be undone by
/// where the phone happens to be lying, and a learner whose phone IS in front
/// of them is exactly the person who can tell us so.
enum MicPreference: String, CaseIterable, Identifiable {
    /// Whatever the learner is wearing (default).
    case earphone
    /// The iPhone's own mic, even with earphones connected. Output stays on
    /// hi-fi A2DP in the earphones — only the input moves.
    case phone

    var id: String { rawValue }
}

/// UserDefaults-backed, global across languages and surfaces: this is a fact
/// about the learner's hardware setup, not about what they're studying.
///
/// Deliberately NOT asked before every session. The answer is the same on the
/// second ask as on the first, and Talk's whole habit story is one tap from
/// tapping to talking — a modal in front of that is a tax on the app's most
/// repeated action. So: ask once, on the first mic use with Bluetooth
/// connected, then live in Me → Voice.
enum MicPreferenceStore {

    static let key = "futurevoice.micPreference"
    private static let askedKey = "futurevoice.micPreference.asked"

    static var current: MicPreference {
        get {
            UserDefaults.standard.string(forKey: key)
                .flatMap(MicPreference.init(rawValue:)) ?? .earphone
        }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: key) }
    }

    /// The one thing the audio layer needs to know.
    static var forcesBuiltInMic: Bool { current == .phone }

    /// Whether the learner has been asked (by the sheet) or has visited the
    /// setting. Either counts — both mean the choice is now theirs.
    static var hasChosen: Bool {
        get { UserDefaults.standard.bool(forKey: askedKey) }
        set { UserDefaults.standard.set(newValue, forKey: askedKey) }
    }

    static func choose(_ preference: MicPreference) {
        current = preference
        hasChosen = true
    }

    /// Ask only when the choice is live: a Bluetooth device is connected AND
    /// we've never asked. Everything else — wired headphones, speaker, a
    /// second session — goes straight through.
    static func shouldAsk() -> Bool {
        !hasChosen && AudioSessionRouting.hasBluetoothOutput()
    }
}

File: android/app/src/main/java/com/roro/futurevoice/data/MicPreference.kt
package com.roro.futurevoice.data

import android.content.Context
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Build

/**
 * Which mic records the learner when earphones are connected.
 *
 * Framed by SITUATION, not hardware, and honest in both directions: the
 * phone's mic really is the better microphone (wideband, multi-mic), and the
 * earphone's really does sound like a phone call. What the earphone buys is
 * POSITION — it is at the mouth wherever the phone ends up. Which matters
 * more depends on a fact the app cannot see, so it asks about that fact once.
 *
 * Asked ONCE, on the first mic use with a Bluetooth headset connected, then
 * it lives in Me → Voice. Not per session: the answer is the same the second
 * time, and Talk's whole story is one tap from tapping to talking.
 *
 * The VOICE CLONE is the exception and never asks — a clone must be built
 * from the best microphone in the room, so that flow forces the built-in one.
 */
object MicPreference {
    const val EARPHONE = "earphone"   // whatever they're wearing (default)
    const val PHONE = "phone"         // the phone's own mic, output stays in the earphones

    private const val PREFS = "futurevoice"
    private const val KEY = "futurevoice.micPreference"
    private const val ASKED = "futurevoice.micPreference.asked"

    fun current(c: Context): String =
        c.getSharedPreferences(PREFS, 0).getString(KEY, EARPHONE) ?: EARPHONE

    /** The one thing the audio layer needs to know. */
    fun forcesBuiltInMic(c: Context) = current(c) == PHONE

    fun set(c: Context, value: String) {
        c.getSharedPreferences(PREFS, 0).edit().putString(KEY, value).putBoolean(ASKED, true).apply()
    }

    fun hasChosen(c: Context) = c.getSharedPreferences(PREFS, 0).getBoolean(ASKED, false)

    /** Only worth asking where the answer changes anything. */
    fun shouldAsk(c: Context): Boolean = !hasChosen(c) && headsetConnected(c)

    fun headsetConnected(c: Context): Boolean {
        val am = c.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return false
        return am.getDevices(AudioManager.GET_DEVICES_INPUTS).any {
            it.type == AudioDeviceInfo.TYPE_BLUETOOTH_SCO ||
                it.type == AudioDeviceInfo.TYPE_BLE_HEADSET ||
                it.type == AudioDeviceInfo.TYPE_WIRED_HEADSET ||
                it.type == AudioDeviceInfo.TYPE_USB_HEADSET
        }
    }

    /**
     * Route the INPUT. Android picks the headset mic on its own for a
     * communication recording, so only the "phone mic" choice needs saying —
     * and saying it must not move the OUTPUT, which stays in the earphones.
     */
    fun applyInputRoute(c: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return
        val am = c.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        if (!forcesBuiltInMic(c)) { runCatching { am.clearCommunicationDevice() }; return }
        val builtIn = am.availableCommunicationDevices
            .firstOrNull { it.type == AudioDeviceInfo.TYPE_BUILTIN_MIC } ?: return
        runCatching { am.setCommunicationDevice(builtIn) }
    }

    /** The clone's own recording: the best mic in the room, never a choice. */
    fun forceBuiltInForClone(c: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return
        val am = c.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        am.availableCommunicationDevices.firstOrNull { it.type == AudioDeviceInfo.TYPE_BUILTIN_MIC }
            ?.let { runCatching { am.setCommunicationDevice(it) } }
    }
}
